src/scores/categorical/contingency_impl.py

```diff
@@ -0,0 +1,263 @@
+        self.fp_count = self.fp.sum()  # Count of false positives
+        self.fn_count = self.fn.sum()  # Count of true negatives
+        self.total_count = self.tp_count + self.tn_count + self.fp_count + self.fn_count
+
+    def generate_counts(self, *, reduce_dims=None, preserve_dims=None):
+        to_reduce = scores.utils.gather_dimensions(
+            self.forecast_events.dims, self.observed_events.dims, reduce_dims=reduce_dims, preserve_dims=preserve_dims
+        )
+
+        cd = {
+            "tp_count": self.tp.sum(to_reduce),
+            "tn_count": self.tn.sum(to_reduce),
+            "fp_count": self.fp.sum(to_reduce),
+            "fn_count": self.fn.sum(to_reduce),
+        }
+        total = cd["tp_count"] + cd["tn_count"] + cd["fp_count"] + cd["fn_count"]
+        cd["total_count"] = total
+
+        return cd
+
+    def accuracy(self, *, preserve_dims=None, reduce_dims=None):
+        """
+        The proportion of forecasts which are true
+        """
+        count_dictionary = self.generate_counts(preserve_dims=preserve_dims, reduce_dims=reduce_dims)
+        correct_count = count_dictionary["tp_count"] + count_dictionary["tn_count"]
+        ratio = correct_count / count_dictionary["total_count"]
+        return ratio
+
+    def frequency_bias(self, *, preserve_dims=None, reduce_dims=None):
+        """ """
+        cd = self.generate_counts(preserve_dims=preserve_dims, reduce_dims=reduce_dims)
+        freq_bias = (cd["tp_count"] + cd["fp_count"]) / (cd["tp_count"] + cd["fn_count"])
+
+        return freq_bias
+
+    def hit_rate(self, *, preserve_dims=None, reduce_dims=None):
+        """
+        What proportion of the observed events where correctly forecast?
+        Identical to probability_of_detection
+        Range: 0 to 1.  Perfect score: 1.
+        """
+        cd = self.generate_counts(preserve_dims=preserve_dims, reduce_dims=reduce_dims)
+        pod = cd["tp_count"] / (cd["tp_count"] + cd["fn_count"])
+
+        return pod
+
+    def probability_of_detection(self, *, preserve_dims=None, reduce_dims=None):
+        """
+        What proportion of the observed events where correctly forecast?
+        Identical to hit_rate
+        Range: 0 to 1.  Perfect score: 1.
+        """
+        cd = self.generate_counts(preserve_dims=preserve_dims, reduce_dims=reduce_dims)
+        pod = cd["tp_count"] / (cd["tp_count"] + cd["fn_count"])
+
+        return pod
+
+    def false_alarm_rate(self, *, preserve_dims=None, reduce_dims=None):
+        """
+        What fraction of the non-events were incorrectly predicted?
+        Identical to probability_of_false_detection
+        Range: 0 to 1.  Perfect score: 0.
+        """
+        cd = self.generate_counts(preserve_dims=preserve_dims, reduce_dims=reduce_dims)
+        far = cd["fp_count"] / (cd["tn_count"] + cd["fp_count"])
+
+        return far
+
+    def probability_of_false_detection(self, *, preserve_dims=None, reduce_dims=None):
```

Comment:
[optional]

Could just pass through to `false_alarm_rate()`

Reply:
Done.